File: interfaces/java/com/kx/Type.java
package com.kx;

import java.lang.reflect.Array;
import java.sql.Date;
import java.sql.Time;
import java.sql.Timestamp;
import java.text.DecimalFormat;

/**
 * This class is a dumping ground for static methods and values from 
 * the legacy implementation.
 * 
 * @see <a href="https://code.kx.com/trac/wiki/Cookbook/InterfacingWithJava">Interfacing With Java at code.kx</a>
 * @author tech@kx.com
 */
public final class Type {

	/* No one is allowed instantiate this. */	
	private Type(){}

	/**
	 * This is the kdb+ null integer value.
	 */
	public final static int NullInteger = Integer.MIN_VALUE;

	/**
	 * This is the kdb+ null long value.
	 */
	public final static long NullLong = Long.MIN_VALUE;

	/**
	 * This is the kdb+ null float value.
	 */
	public final static double NullFloat = Double.NaN;

	/**
	 * Gets the appropriate null value for the given type code.
	 * 
	 * @param c the type identifier
	 * @return the null value
	 */
	public static Object getNull(char c) {
		return NULLVALUES[" b  xhijefcs mdz uvt".indexOf(c)];
	}

	/**
	 * Formats an integer to fit in a 2 digit string.
	 * 
	 * @param i the integer to format
	 * @return the 2 digit formatted string
	 */
	public static String toTwoDigitString(int i) {
		return new DecimalFormat("00").format(i);
	}

	/**
	 * Given an object, what is the appropriate kdb+ type tag?
	 * The kdb+ connection is that negative type tags are scalar values
	 * while positive ones are vectors. There are a number of special 
	 * cases. Notably, Flips, Dicts, and general lists.
	 *  
	 * @param x some object
	 * @return the appropriate kdb+ type tag
	 */
	public static int typeOf(Object x) {
		return x instanceof Boolean ? -1 : x instanceof Byte ? -4
			: x instanceof Short ? -5 : x instanceof Integer ? -6
			: x instanceof Long ? -7 : x instanceof Float ? -8
			: x instanceof Double ? -9 : x instanceof Character ? -10
			: x instanceof String ? -11 : x instanceof Month ? -13
			: x instanceof Time ? -19 : x instanceof Date ? -14
			: x instanceof java.util.Date ? -15 : x instanceof Minute ? -17
			: x instanceof Second ? -18 : x instanceof boolean[] ? 1
			: x instanceof byte[] ? 4 : x instanceof short[] ? 5
			: x instanceof int[] ? 6 : x instanceof long[] ? 7
			: x instanceof float[] ? 8 : x instanceof double[] ? 9
			: x instanceof char[] ? 10 : x instanceof String[] ? 11
			: x instanceof Month[] ? 13 : x instanceof Time[] ? 19
			: x instanceof Date[] ? 14 : x instanceof java.util.Date[] ? 15
			: x instanceof Minute[] ? 17 : x instanceof Second[] ? 18
			: x instanceof Flip ? 98 : x instanceof Dict ? 99
			: 0;
	}

	/**
	 * What is the length of the kdb+ object?
	 * 
	 * @param x some kdb+ value
	 * @return the size
	 */
	public static int size(Object x) {
		return x instanceof Dict ? size(((Dict) x).getKeys())
			: x instanceof Flip ? size(((Flip) x).getColumnValues()[0]) : Array.getLength(x);
	}

	/**
	 * Converts an object to a Flip.
	 * 
	 * @param object
	 * @return the flip
	 */
	public static Flip toFlip(Object object) {
		if (typeOf(object) == 98)
			return (Flip) object;
		Dict d = (Dict) object;
		Flip a = (Flip) d.getKeys(), b = (Flip) d.getValues();
		int m = size(a.getColumnNames()), n = size(b.getColumnNames());
		String[] names = new String[m + n];
		System.arraycopy(a.getColumnNames(), 0, names, 0, m);
		System.arraycopy(b.getColumnNames(), 0, names, m, n);
		Object[] values = new Object[m + n];
		System.arraycopy(a.getColumnValues(), 0, values, 0, m);
		System.arraycopy(b.getColumnValues(), 0, values, m, n);
		return new Flip(new Dict(names, values));
	}

	/**
	 * What is the kdb+ length of this string?
	 * 
	 * @param string
	 * @return the length
	 */
	static int stringLength(String string) {
		int i;
		return string == null 
					? 0 
					: -1 < (i = string.indexOf('\000')) ? i : string.length();
	}

	/**
	 * Get the i'th value in the given kdb+ value.
	 * 
	 * @param x some kdb+ object
	 * @param i the index
	 * @return the value at the given index
	 */
	public static Object at(Object x, int i) {
		Object target = Array.get(x, i);
		int t = -typeOf(target);
		boolean isNull = t > 4 && target.equals(NULLVALUES[t]);
		return isNull ? null : target;
	}

	/**
	 * Sets x[i] to y.
	 * 
	 * @param x the collection
	 * @param i the index
	 * @param y the new value
	 */
	public static void set(Object x, int i, Object y) {
		Array.set(x, i, null == y ? NULLVALUES[typeOf(x)] : y);
	}

	/**
	 * Computes the object size.
	 * 
	 * @param o some kdb+ value
	 * @return the size
	 */
	public static int objectSize(Object o) {
		int i = 0, n, t = typeOf(o), j;
		if (t == 99)
			return 1 + objectSize(((Dict) o).getKeys()) + objectSize(((Dict) o).getValues());
		if (t == 98)
			return 3 + objectSize(((Flip) o).getColumnNames()) + objectSize(((Flip) o).getColumnValues());
		if (t < 0)
			return t == -11 ? 2 + stringLength((String) o) : 1 + basesize[-t];
		j = 6;
		n = size(o);
		if (t == 0 || t == 11)
			for (; i < n; ++i)
				j += t == 0 ? objectSize(((Object[]) o)[i]) : 1 + stringLength(((String[]) o)[i]);
		else
			j += n * basesize[t];
		return j;
	}

	private static int[] basesize = 
		{0, 1, 0, 0, 1, 2, 4, 8, 4, 8, 1, 0, 0, 4, 4, 8, 0, 4, 4, 4 };

	private static Object[] NULLVALUES = { 
			null, 
			new Boolean(false), 
			null, 
			null,
			new Byte((byte) 0), 
			new Short(Short.MIN_VALUE), 
			new Integer(NullInteger),
			new Long(NullLong), 
			new Float(NullFloat), 
			new Double(NullFloat), 
			new Character(' '),
			"", 
			null, 
			new Month(NullInteger), 
			new Date(NullLong), 
			new Timestamp(NullLong), 
			null,
			new Minute(NullInteger), 
			new Second(NullInteger), 
			new Time(NullLong) 
	};

}

File: interfaces/csharp/Kdbplus/Type.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Kdbplus
{
    /// <summary>
    /// This class is a dumping ground for static methods and values from 
    /// the legacy implementation.
    /// </summary>
    public class Type
    {
        /// <summary>
        /// This is the kdb+ null integer value.
        /// </summary>
        public static int NullInteger = Int32.MinValue;

        /// <summary>
        /// This is the kdb+ null long value.
        /// </summary>
        public static long NullLong = Int64.MinValue;

        /// <summary>
        /// This is the kdb+ null float value.
        /// </summary>
        public static double NullFloat = Double.NaN;

        /// <summary>
        ///  Gets the appropriate null value for the given type code.
        /// </summary>
        /// <param name="code">The kdb+ type code.</param>
        /// <returns>The appropriate null value.</returns>
        public static object GetNull(char code)
        {
            return nullValues[" b  xhijefcs mdz uvt".IndexOf(code)];
        }

        /// <summary>
        /// What is the kdb+ length of the string?
        /// </summary>
        /// <param name="s">The string to compute the length of.</param>
        /// <returns>The length of the string.</returns>
        public static int StringLength(string s)
        {
            var i = s.IndexOf('\0');
            return -1 < i ? i : s.Length;
        }

        /// <summary>
        /// Formats an integer to fit in a 2 digit string
        /// </summary>
        /// <param name="i">The integer to format.</param>
        /// <returns>The 2 digit formatted string.</returns>
        public static string ToTwoDigitString(int i)
        {
            return String.Format("{0:00}", i);
        }

        /// <summary>
        /// Check whether this is a null value.
        /// </summary>
        /// <param name="x">The object to check.</param>
        /// <returns>Whether this is null or not.</returns>
        public static bool IsNull(object x)
        {
            var type = -Type.TypeOf(x);
            return type > 4 && x.Equals(nullValues[type]);
        }

        /// <summary>
        /// Converts an object to a Flip.
        /// </summary>
        /// <param name="obj">The object from which to build the Flip.</param>
        /// <returns>The flipped object.</returns>
        public static Flip ToFlip(object obj)
        {
            if (TypeOf(obj) == 98) 
                return (Flip) obj;
            var d = (Dict) obj;
            Flip a = (Flip) d.Keys, b = (Flip) d.Values;
            var m = Type.Length(a.TheColumnNames);
            var n = Type.Length(b.TheColumnNames);
            var x = new string[m + n];
            Array.Copy(a.TheColumnNames, 0, x, 0, m);
            Array.Copy(b.TheColumnNames, 0, x, m, n);
            var y = new object[m + n];
            Array.Copy(a.TheColumnValues, 0, y, 0, m);
            Array.Copy(b.TheColumnValues, 0, y, m, n);
            return new Flip(new Dict(x, y));
        }

        /// <summary>
        /// Check the type of the given kdb+ value.
        /// </summary>
        /// <param name="obj">The object to type check.</param>
        /// <returns>The kdb+ type for the object.</returns>
        public static int TypeOf(object obj)
        {
            return obj is bool ? -1
                : obj is byte ? -4
                : obj is short ? -5
                : obj is int ? -6
                : obj is long ? -7
                : obj is float ? -8
                : obj is double ? -9
                : obj is char ? -10
                : obj is string ? -11
                : obj is Month ? -13
                : obj is Date ? -14
                : obj is DateTime ? -15
                : obj is Minute ? -17
                : obj is Second ? -18
                : obj is TimeSpan ? -19
                : obj is bool[] ? 1
                : obj is byte[] ? 4
                : obj is short[] ? 5
                : obj is int []? 6 
                : obj is long[] ? 7
                : obj is float[] ? 8
                : obj is double [] ? 9
                : obj is char[] ? 10
                : obj is DateTime[] ? 15
                : obj is TimeSpan[] ? 19
                : obj is Flip ? 98
                : obj is Dict ? 99
                : 0;
        }

        /// <summary>
        /// Compute the length of the given object.
        /// </summary>
        /// <param name="obj">The kdb+ object of which to calculate the length.</param>
        /// <returns>The length.</returns>
        public static int Length(object obj)
        {
            return obj is Dict ? Length(((Dict) obj).Keys) : obj is Flip ? Length(((Flip) obj).TheColumnValues[0]) : ((Array) obj).Length;
        }

        /// <summary>
        /// Computes the object Length.
        /// </summary>
        /// <param name="x">The object of which to compute the Length.</param>
        /// <returns>The Length of the object.</returns>
        public static int ObjectSize(object x)
        {
            var type = Type.TypeOf(x);
            if (type == 99) 
                return 1 + ObjectSize(((Dict) x).Keys) + ObjectSize(((Dict) x).Values);
            if (type == 98) 
                return 3 + ObjectSize(((Flip) x).TheColumnNames) + ObjectSize(((Flip) x).TheColumnValues);
            if (type < 0) 
                return type == -11 ? 2 + Type.StringLength((string) x) : 1 + basesize[-type];
            var j = 6;
            var length = Type.Length(x);
            if (type == 0) 
                for (var i = 0; i < length; ++i) 
                    j += ObjectSize(((object[]) x)[i]);
            else j += length*basesize[type];
            return j;
        }

        /// <summary>
        /// Get the i'th value in the given kdb+ value.
        /// </summary>
        /// <param name="x">Some kdb+ value.</param>
        /// <param name="i">The position At which to index.</param>
        /// <returns>The value At x[i]</returns>
        public static object At(object x, int i)
        {
            var obj = ((Array) x).GetValue(i);
            return IsNull(obj) ? null : obj;
        }

        /// <summary>
        /// User for date offset calculations in kdb+.
        /// </summary>
        public static long offset = (long) 8.64e11*730119;

        private static readonly object[] nullValues = 
        {
              null, false, null, null, (byte) 0, Int16.MinValue, NullInteger, NullLong, 
              (Single) NullFloat, NullFloat, ' ', "", null, new Month(NullInteger), new Date(NullInteger), 
              new DateTime(0), null, new Minute(NullInteger),new Second(NullInteger), 
              new TimeSpan(NullLong) 
        };

        private static readonly int[] basesize = { 0, 1, 0, 0, 1, 2, 4, 8, 4, 8, 1, 0, 0, 4, 4, 8, 4, 4, 4, 4 };
      
    }
}
